cmd/kubeadm/app/util/endpoint_test.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"testing"
+)
+
+func TestGenerateMasterEndpoint(t *testing.T) {
+	var tests = []struct {
+		address  string
+		port     int32
+		endpoint string
+		expected bool
+	}{
+		{
+			address:  "1.2.3.4",
+			port:     6443,
+			endpoint: "https://1.2.3.4:6443",
+			expected: true,
+		},
+		{
+			address:  "2001:db8::4",
+			port:     6443,
+			endpoint: "https://[2001:db8::4]:6443",
+			expected: true,
+		},
+		{
+			address:  "1.2.3.4",
+			port:     6443,
+			endpoint: "https://[1.2.3.4]:6443",
+			expected: false,
+		},
+		{
+			address:  "2001:db8::4",
+			port:     6443,
+			endpoint: "https://2001:db8::4:6443",
+			expected: false,
+		},
+	}
+	for _, rt := range tests {
+		actual, err := GenerateMasterEndpoint(rt.address, rt.port)
+		if err != nil && !rt.expected {
```

Comment:
So instead of having rt.expected be a boolean, it could be of type string, and could be tested against err to see if the error is the same (especially if non-nil). Something like...

```
if err != nil {
    if !strings.Contains(err.Error(), tc.errStrFrag) {
        ... wrong error occurred
    }
} else {
    if tc.errStrFrag != nil {
        ... should have been an error
    } else if actual != rt.endpoint {
        ... incorrect value
    }
}
```